File: src/backend/api/Features/Exercises/ExerciseTypes.cs
namespace SalesTrainer.Api.Features.Exercises;

/// <summary>
/// All supported exercise type identifiers.
/// Change values here to rename types across the entire backend.
/// </summary>
public static class ExerciseTypes
{
    public const string ChooseOption = "choose_option";
    public const string FillBlank = "fill_blank";
    public const string Reorder = "reorder";
    public const string MatchPairs = "match_pairs";
    public const string Categorize = "categorize";
    public const string SpotMistake = "spot_mistake";
    public const string Rewrite = "rewrite";
    public const string AiDialogue = "ai_dialogue";
    public const string EvaluateCall = "evaluate_call";
    public const string FreeText = "free_text";

    /// <summary>
    /// Theory card — a non-graded "story" card the learner swipes through. A lesson
    /// made entirely of theory_card exercises is a theory lesson (see NEW_EXERCISE_TYPES.md).
    /// </summary>
    public const string TheoryCard = "theory_card";

    public static readonly string[] All =
    [
        ChooseOption,
        FillBlank,
        Reorder,
        MatchPairs,
        Categorize,
        SpotMistake,
        Rewrite,
        AiDialogue,
        EvaluateCall,
        FreeText,
        TheoryCard
    ];

    public static readonly string[] AiPowered =
    [
        SpotMistake,
        Rewrite,
        AiDialogue,
        EvaluateCall,
        FreeText
    ];
}

File: src/frontend/features/exercise/types/exercise-types.ts
/**
 * All supported exercise type identifiers.
 * Change values here to rename types across the entire frontend.
 */
export const ExerciseTypes = {
  ChooseOption: 'choose_option',
  FillBlank: 'fill_blank',
  Reorder: 'reorder',
  MatchPairs: 'match_pairs',
  Categorize: 'categorize',
  SpotMistake: 'spot_mistake',
  Rewrite: 'rewrite',
  AiDialogue: 'ai_dialogue',
  EvaluateCall: 'evaluate_call',
  FreeText: 'free_text',
  TheoryCard: 'theory_card',
} as const;

export type ExerciseType = typeof ExerciseTypes[keyof typeof ExerciseTypes];

export const AllExerciseTypes = Object.values(ExerciseTypes);

export const AiPoweredExerciseTypes: ExerciseType[] = [
  ExerciseTypes.SpotMistake,
  ExerciseTypes.Rewrite,
  ExerciseTypes.AiDialogue,
  ExerciseTypes.EvaluateCall,
  ExerciseTypes.FreeText,
];

export const isAiPoweredExercise = (type: ExerciseType): boolean =>
  AiPoweredExerciseTypes.includes(type);
